File: 743. Network Delay Time/743. Network Delay Time.js
/**
 * @param {number[][]} times
 * @param {number} n
 * @param {number} k
 * @return {number}
 */
var networkDelayTime = function (times, n, k) {
  const adjList = new Map();
  for (const [source, target, time] of times) {
    if (!adjList.has(source)) {
      adjList.set(source, []);
    }
    adjList.get(source).push([target, time]);
  }

  let delay = 0;

  const minHeap = new MinPriorityQueue({ compare: (a, b) => a[0] - b[0] });
  minHeap.enqueue([0, k]);

  const visited = new Set();

  while (!minHeap.isEmpty()) {
    const [time, node] = minHeap.dequeue();

    if (visited.has(node)) {
      continue;
    }

    visited.add(node);
    delay = Math.max(time, delay);

    if (adjList.has(node)) {
      for (const [node2, time2] of adjList.get(node)) {
        if (!visited.has(node2)) {
          minHeap.enqueue([time + time2, node2]);
        }
      }
    }
  }

  return visited.size === n ? delay : -1;
};

File: 743. Network Delay Time.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        adj_list = defaultdict(list)

        for u, v, time in times:
            adj_list[u].append((v, time))

        min_times = {}
        min_heap = [(0, k)]

        while min_heap:
            time, node = heapq.heappop(min_heap)

            if node in min_times:
                continue

            min_times[node] = time
            for nei, nei_time in adj_list[node]:
                if nei not in min_times:
                    heapq.heappush(min_heap, (nei_time + time, nei))

        return max(min_times.values()) if len(min_times) == n else -1
